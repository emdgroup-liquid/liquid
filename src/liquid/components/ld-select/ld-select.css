:host {
  display: inline-flex;

  .ld-select {
    flex-grow: 1;
    max-width: 100%;
  }
}

:host,
.ld-select {
  /* layout */
  --ld-select-min-width: 12.8125rem;
  --ld-select-min-height-sm: 2rem;
  --ld-select-min-height-md: 2.5rem;
  --ld-select-min-height-lg: 3.125rem;
  --ld-select-padding-x-sm: 0.625rem;
  --ld-select-padding-x-md: 0.75rem;
  --ld-select-padding-x-lg: 0.875rem;
  --ld-select-padding-y-sm: 0.25rem;
  --ld-select-padding-y-md: 0.5rem;
  --ld-select-padding-y-lg: 0.5rem;
  --ld-select-padding-right-sm: calc(2.5rem + 0.625rem);
  --ld-select-padding-right-md: calc(2.5rem + 0.75rem);
  --ld-select-padding-right-lg: calc(2.5rem + 0.875rem);
  --ld-select-icon-size-sm: 1rem;
  --ld-select-icon-size-md: 1.25rem;
  --ld-select-icon-size-lg: 1.5rem;
  --ld-select-trigger-line-height: 1.25;

  /* colors */
  --ld-select-col: var(--ld-col-neutral-900);
  --ld-select-col-disabled: var(--ld-col-neutral-100);
  --ld-select-col-border: var(--ld-col-neutral-100);
  --ld-select-col-border-hover: var(--ld-col-neutral-300);
  --ld-select-bg-col: var(--ld-col-wht);
  --ld-select-selection-col: var(--ld-col-wht);

  /* themable colors */
  --ld-select-invalid-col: var(--ld-thm-error);
  --ld-select-invalid-icon-col-hover: var(--ld-thm-error-hover);
  --ld-select-invalid-icon-col-focus: var(--ld-thm-error-focus);
  --ld-select-invalid-icon-col-active: var(--ld-thm-error-active);
  --ld-select-invalid-disabled-bg-col: var(--ld-thm-error-disabled);
  --ld-select-thm-col: var(--ld-thm-primary);
  --ld-select-thm-col-hover: var(--ld-thm-primary-hover);
  --ld-select-thm-col-focus: var(--ld-thm-primary-focus);
  --ld-select-thm-col-active: var(--ld-thm-primary-active);
  --ld-select-ghost-trigger-bg-col-hover: var(--ld-thm-primary-alpha-lowest);
  --ld-select-ghost-trigger-bg-col-focus: var(--ld-thm-primary-alpha-low);

  /* reset */
  *,
  *::before,
  *::after {
    box-sizing: border-box;
  }
  ul {
    list-style: none;
  }
}

.ld-select {
  display: inline-flex;
  position: relative;

  select + .ld-icon,
  ::slotted(ld-icon) {
    width: var(--ld-select-icon-size-md);
    height: var(--ld-select-icon-size-md);
  }

  ::slotted(ld-icon) {
    display: contents !important;
  }

  select {
    + .ld-icon,
    + .ld-select__icon {
      position: absolute;
      right: var(--ld-select-padding-x-md);
      top: 50%;
      transform: translateY(-50%);
    }

    + .ld-icon {
      width: var(--ld-select-icon-size-md);
      height: var(--ld-select-icon-size-md);
    }

    &:where(:not[multiple]) {
      padding-right: var(--ld-select-padding-right-md);
    }
  }

  &.ld-select--sm {
    select {
      + .ld-icon,
      + .ld-select__icon {
        right: var(--ld-select-padding-x-sm);
      }

      + .ld-icon {
        width: var(--ld-select-icon-size-sm);
        height: var(--ld-select-icon-size-sm);
      }

      &:where(:not[multiple]) {
        padding-right: var(--ld-select-padding-right-sm);
      }
    }
  }
  &.ld-select--lg {
    select {
      + .ld-icon,
      + .ld-select__icon {
        right: var(--ld-select-padding-x-lg);
      }

      + .ld-icon {
        width: var(--ld-select-icon-size-lg);
        height: var(--ld-select-icon-size-lg);
      }

      &:where(:not[multiple]) {
        padding-right: var(--ld-select-padding-right-lg);
      }
    }
  }
}

.ld-select--expanded {
  .ld-tether-target-attached-bottom {
    .ld-select__btn-trigger:not(.ld-select__btn-trigger--detached) {
      border-bottom-left-radius: 0;
      border-bottom-right-radius: 0;
    }
  }
}

:where(.ld-select:not(.ld-select--inline, .ld-select--ghost)) {
  min-width: var(--ld-select-min-width);
}

.ld-select__select,
.ld-select__btn-trigger {
  width: 100%;
}

.ld-select > select,
.ld-select__btn-trigger {
  /* outline: none; */
  position: relative;
  display: flex;
  align-items: center;
  justify-content: flex-end;
  font: var(--ld-typo-body-m);
  line-height: var(--ld-select-trigger-line-height);
  border: 0;
  padding: var(--ld-select-padding-y-md) var(--ld-select-padding-x-md);
  border-radius: var(--ld-br-m);
  height: 100%;
  user-select: none;
  touch-action: manipulation;
  color: var(--ld-select-col);
  background-color: var(--ld-select-bg-col);
  text-align: left;
  appearance: none;
  -webkit-touch-callout: none;

  &:where(select) {
    width: 100%;

    &[multiple] {
      optgroup {
        padding: var(--ld-sp-8) 0;
        font-style: normal;
      }

      option {
        padding: var(--ld-sp-6) var(--ld-sp-6);
        margin: 0 calc(-1 * var(--ld-sp-6));
        border-radius: var(--ld-br-m);

        &:not(:last-of-type) {
          margin-bottom: var(--ld-sp-2);
        }
      }
    }
  }

  &:where(select:not(:disabled, [aria-disabled='true'])),
  &:where(.ld-select__btn-trigger:not([aria-disabled='true'])) {
    cursor: pointer;
  }

  &:where([aria-expanded='true']) {
    z-index: 2;

    &:not(:focus:focus-visible, .ld-select__btn-trigger--detached) {
      box-shadow: none;
    }
  }

  &:disabled,
  &:disabled + .ld-select__icon,
  &:disabled + .ld-icon,
  &:where([aria-disabled='true']),
  &:where([aria-disabled='true']) + .ld-select__icon,
  &:where([aria-disabled='true']) + .ld-icon {
    color: var(--ld-select-col-disabled);
  }
}

.ld-select__btn-trigger {
  overflow: hidden;
}

:where(.ld-select),
:where(.ld-select) > select {
  min-height: var(--ld-select-min-height-md);
}

:where(.ld-select--sm),
:where(.ld-select--sm) > select {
  min-height: var(--ld-select-min-height-sm);
}

:where(.ld-select--lg),
:where(.ld-select--lg) > select {
  min-height: var(--ld-select-min-height-lg);
}

.ld-select--sm > select,
.ld-select--sm .ld-select__btn-trigger {
  padding: var(--ld-select-padding-y-sm) var(--ld-select-padding-x-sm);
}

.ld-select--lg > select,
.ld-select--lg .ld-select__btn-trigger {
  padding: var(--ld-select-padding-y-lg) var(--ld-select-padding-x-lg);
}

.ld-select--ghost {
  &:not(ld-select),
  select,
  .ld-select__btn-trigger {
    background-color: transparent;
  }

  select,
  .ld-select__btn-trigger {
    &:not(:focus) {
      box-shadow: none;
    }
  }
}

.ld-select:where(:not(.ld-select--ghost)) {
  &:where(.ld-select--detached),
  &:where(:not(.ld-select--detached, .ld-select--expanded)) {
    :where(select),
    :where(.ld-select__btn-trigger) {
      box-shadow: inset 0 0 0 var(--ld-sp-2) var(--ld-select-col-border);
    }
  }

  &:where(:not(.ld-select--invalid)) {
    select,
    .ld-select__btn-trigger {
      &:where(
          .ld-select__btn-trigger--detached:not(
              :disabled,
              [aria-disabled='true']
            )
        ),
      &:where(
          :not(
              :disabled,
              [aria-disabled='true'],
              .ld-select__btn-trigger--detached,
              [aria-expanded='true']
            )
        ) {
        @media (hover: hover) {
          &:hover:not(:focus:focus-visible) {
            box-shadow: inset 0 0 0 var(--ld-sp-2)
              var(--ld-select-col-border-hover);
          }
        }
      }
    }
  }
}

.ld-select--invalid > select,
.ld-select--invalid .ld-select__btn-trigger--invalid {
  &:where(:not(:disabled)) {
    &:where(:focus:focus-visible) {
      box-shadow: inset 0 0 0 var(--ld-sp-2) var(--ld-select-invalid-col);
    }
  }

  &:where(:not(:disabled, [aria-disabled='true'])) {
    &:where(:not(:focus:focus-visible)) {
      background-color: var(--ld-select-invalid-disabled-bg-col);
      color: var(--ld-select-invalid-col);
    }

    &:where(:not(.ld-select__btn-trigger--ghost)) {
      box-shadow: inset 0 0 0 var(--ld-sp-2) var(--ld-select-invalid-col);
    }

    + .ld-icon,
    + .ld-select__icon,
    .ld-select__icon {
      color: var(--ld-select-invalid-col);
    }

    @media (hover: hover) {
      &:hover {
        box-shadow: inset 0 0 0 var(--ld-sp-2) var(--ld-select-invalid-col);

        + .ld-icon,
        + .ld-select__icon,
        .ld-select__icon,
        ::slotted(ld-icon) {
          color: var(--ld-select-invalid-icon-col-hover);
        }
      }
    }
    &:where(:focus:focus-visible) {
      + .ld-icon,
      + .ld-select__icon,
      .ld-select__icon,
      ::slotted(ld-icon) {
        color: var(--ld-select-invalid-icon-col-focus);
      }
    }
    &:where(:active),
    &:where(:active:focus-visible) {
      + .ld-icon,
      + .ld-select__icon,
      .ld-select__icon,
      ::slotted(ld-icon) {
        color: var(--ld-select-invalid-icon-col-active);
      }
    }
  }
}

.ld-select__btn-trigger-text-wrapper,
.ld-select__selection-list {
  flex-grow: 1;
}

.ld-select__btn-trigger-text-wrapper {
  font: var(--ld-typo-label-m);
  line-height: var(--ld-select-trigger-line-height);
  padding: var(--ld-sp-6) var(--ld-sp-8) var(--ld-sp-6) 0;
  height: calc(100% + var(--ld-sp-12));
  margin: calc(-1 * var(--ld-sp-6)) 0;
  display: flex;
  align-items: center;
  overflow: hidden;
}

.ld-select--sm > select,
.ld-select--sm .ld-select__btn-trigger-input,
.ld-select--sm .ld-select__btn-trigger-text-wrapper {
  font: var(--ld-typo-body-s);
  line-height: var(--ld-select-trigger-line-height);
}
.ld-select--lg > select,
.ld-select--lg .ld-select__btn-trigger-input,
.ld-select--lg .ld-select__btn-trigger-text-wrapper {
  font: var(--ld-typo-body-l);
  line-height: var(--ld-select-trigger-line-height);
}

.ld-select > select,
.ld-select__btn-trigger-text,
.ld-select__selection-label-text {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.ld-select__selection-list-container {
  display: flex;
  flex-direction: column-reverse;
  gap: var(--ld-sp-6);
  margin-right: auto;
}

.ld-select__selection-list {
  display: flex;
  flex-wrap: wrap;
  margin: calc(-1 * var(--ld-sp-1)) var(--ld-sp-4) calc(-1 * var(--ld-sp-4)) 0;
  overflow: hidden;
  padding: 0;
}

.ld-select__selection-list-item {
  flex: 0 1;
  margin-right: var(--ld-sp-4);
  margin-bottom: var(--ld-sp-4);
  width: 100%;
}

.ld-select__selection-list-item--overflowing {
  display: none;
}

.ld-select__selection-list-more {
  order: 2147483647; /* Highest possible */
  display: inline-flex;
  align-items: center;
  font: var(--ld-typo-label-s);
  font-weight: 700;
  padding: var(--ld-sp-4) var(--ld-sp-6);
  border-radius: var(--ld-br-m);
  margin-right: var(--ld-sp-4);
  margin-bottom: var(--ld-sp-4);
}

.ld-select__btn-clear-single,
.ld-select__btn-clear {
  /* outline: none; */
  border: 0;
  padding: 0;
  border-radius: var(--ld-br-full);
  user-select: none;
  touch-action: manipulation;
  background-color: transparent;
  line-height: 0;
  -webkit-touch-callout: none;

  &:not(:disabled) {
    cursor: pointer;
  }
}

.ld-select__btn-clear {
  margin-right: var(--ld-sp-6);
  flex-shrink: 0;

  &:disabled {
    color: var(--ld-select-col-disabled);
  }
}

.ld-select__btn-clear-single {
  z-index: 1;
  margin-left: var(--ld-sp-6);
}

.ld-select__selection-label {
  position: relative;
  width: 100%;
  display: inline-flex;
  align-items: center;
  color: var(--ld-select-selection-col);
  font: var(--ld-typo-label-s);
  font-weight: 700;
  padding: var(--ld-sp-4) var(--ld-sp-6);
}

.ld-select__selection-label-bg {
  position: absolute;
  inset: 0;
  border-radius: var(--ld-br-m);

  :where(.ld-select__btn-trigger[aria-disabled='true']) & {
    background-color: var(--ld-select-col-disabled);
  }
}

.ld-select__selection-label-text {
  z-index: 1;
}

.ld-select__btn-clear-single-icon {
  --ld-select-btn-clear-single-size: 0.75rem;
  width: var(--ld-select-btn-clear-single-size);
  height: var(--ld-select-btn-clear-single-size);
}

.ld-select__btn-clear-icon {
  --ld-select-btn-clear-size: 1.25rem;
  width: var(--ld-select-btn-clear-size);
  height: var(--ld-select-btn-clear-size);
}

.ld-select__icon {
  fill: none;
  width: var(--ld-sp-16);
  height: var(--ld-sp-16);
}

.ld-select select + .ld-icon,
.ld-select__icon,
.ld-select ::slotted(ld-icon) {
  flex-shrink: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  pointer-events: none;

  :where(select[aria-disabled='true']) &,
  :where(select:disabled) &,
  :where(.ld-select__btn-trigger[aria-disabled='true']) & {
    color: var(--ld-select-col-disabled);
  }

  .ld-icon,
  svg {
    width: 100%;
    height: 100%;
  }
}

.ld-select__icon--rotated {
  transform-origin: center;
  transform: rotate(180deg);
}

.ld-select__slot-container {
  display: none;
}

:where(
    .ld-select:not(.ld-select--disabled, [aria-disabled], .ld-select--invalid)
  ) {
  .ld-select__btn-trigger {
    &:where(:focus:focus-visible) {
      box-shadow: inset 0 0 0 var(--ld-sp-2) var(--ld-select-thm-col);
    }
  }

  &.ld-select--ghost {
    :where(select:not(:disabled)),
    :where(.ld-select__btn-trigger) {
      @media (hover: hover) {
        &:hover:not(:focus) {
          background-color: var(--ld-select-ghost-trigger-bg-col-hover);
        }
      }
      &:active,
      &:active:focus-visible {
        background-color: var(--ld-select-ghost-trigger-bg-col-focus);
      }
    }
  }

  :where(select:not(:disabled)),
  :where(.ld-select__btn-trigger) {
    .ld-select__icon,
    + .ld-icon,
    + .ld-select__icon,
    ::slotted(ld-icon) {
      color: var(--ld-select-thm-col);
    }

    @media (hover: hover) {
      &:where(:hover) {
        .ld-select__icon,
        + .ld-icon,
        + .ld-select__icon,
        ::slotted(ld-icon) {
          color: var(--ld-select-thm-col-hover);
        }
      }
    }
    &:where(:focus:focus-visible) {
      box-shadow: inset 0 0 0 var(--ld-sp-2) var(--ld-select-thm-col);

      .ld-select__icon,
      + .ld-select__icon,
      ::slotted(ld-icon) {
        color: var(--ld-select-thm-col-focus);
      }
    }
    &:where(:active),
    &:where(:active:focus-visible) {
      .ld-select__icon,
      + .ld-icon,
      + .ld-select__icon,
      ::slotted(ld-icon) {
        color: var(--ld-select-thm-col-active);
      }
    }
  }
}

:where(.ld-select:not(.ld-select--disabled)):not(.ld-select--invalid) {
  :where(select:not(:disabled)),
  :where(.ld-select__btn-trigger) {
    &:where(:focus:focus-visible) {
      box-shadow: inset 0 0 0 var(--ld-sp-2) var(--ld-select-thm-col);
    }
  }
}

:where(
    .ld-select:not(
        .ld-select--disabled,
        [aria-disabled],
        [aria-disabled='true']
      )
  ) {
  .ld-select__selection-list-more {
    color: var(--ld-select-thm-col);
  }
  .ld-select__selection-label-bg {
    background-color: var(--ld-select-thm-col);
  }

  .ld-select__btn-clear-single {
    &:where(:focus:focus-visible) + .ld-select__selection-label-bg {
      background-color: var(--ld-select-thm-col-focus);
    }

    &:where(:active),
    &:where(:active:focus-visible) {
      + .ld-select__selection-label-bg {
        background-color: var(--ld-select-thm-col-active);
      }
    }
  }

  .ld-select__btn-clear {
    color: var(--ld-select-thm-col);

    &:where(:focus:focus-visible) {
      color: var(--ld-select-thm-col-focus);
    }
    @media (hover: hover) {
      &:where(:hover) {
        color: var(--ld-select-thm-col-hover);
      }
    }
    &:where(:active),
    &:where(:active:focus-visible) {
      color: var(--ld-select-thm-col-active);
    }
  }
}
