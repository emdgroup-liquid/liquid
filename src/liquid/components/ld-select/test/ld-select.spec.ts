import MutationObserver from 'mutation-observer'
import { newSpecPage } from '@stencil/core/testing'
import { LdSelect } from '../ld-select'

global.MutationObserver = MutationObserver

async function triggerPopper(page) {
  const ldSelect = page.root
  const triggerButton = ldSelect.querySelector('.ld-select__btn-trigger')
  await triggerButton.dispatchEvent(new Event('click'))
  await page.waitForChanges()
  await new Promise((resolve) => setTimeout(resolve))
  await page.waitForChanges()
}

describe('ld-select', () => {
  it('renders popper element with copies of slotted options', async () => {
    const page = await newSpecPage({
      components: [LdSelect],
      html: `
        <ld-select placeholder="Pick a fruit" name="fruit">
          <ld-option value="apple">Apple</ld-option>
          <ld-option value="banana">Banana</ld-option>
        </ld-select>
      `,
    })
    await new Promise((resolve) => setTimeout(resolve))

    const ldSelect = page.root
    const triggerButton = ldSelect.querySelector('.ld-select__btn-trigger')

    expect(ldSelect.classList.contains('ld-select--expanded')).toBeFalsy()
    expect(triggerButton.getAttribute('aria-expanded')).toEqual('false')

    await triggerButton.dispatchEvent(new Event('click'))
    await page.waitForChanges()
    await new Promise((resolve) => setTimeout(resolve))

    const body = page.body
    const popper = body.querySelector('.ld-select__popper')
    const slottedOptions = ldSelect.querySelectorAll('ld-option')
    const internalOptions = popper.querySelectorAll('ld-option-internal')

    expect(ldSelect.classList.contains('ld-select--expanded')).toBeTruthy()
    expect(slottedOptions.length).toEqual(2)
    expect(internalOptions.length).toEqual(2)
    expect(triggerButton.getAttribute('aria-expanded')).toEqual('true')
    expect(
      popper.classList.contains('ld-select__popper--expanded')
    ).toBeTruthy()
  })

  it('throws if no ld-option(s) are passed to the default slot', async () => {
    expect.assertions(1)
    try {
      const page = await newSpecPage({
        components: [LdSelect],
        html: '<ld-select placeholder="Pick a fruit" name="fruit"></ld-select>',
      })

      await triggerPopper(page)
    } catch (err) {
      expect(err).toStrictEqual(
        TypeError(
          'ld-select requires at least one ld-option element as a child, but found none.'
        )
      )
    }
  })

  // it('throws if something other than an ld-option is passed to the default slot', async () => {
  //   expect.assertions(1)
  //   try {
  //     const page = await newSpecPage({
  //       components: [LdSelect],
  //       html: `
  //         <ld-select placeholder="Pick a fruit" name="fruit">
  //           <ld-option value="apple">Apple</ld-option>
  //           <option value="banana">Banana</option>
  //         </ld-select>
  //       `,
  //     })
  //
  //     await triggerPopper(page)
  //   } catch (err) {
  //     expect(err).toStrictEqual(
  //       TypeError(
  //         'ld-select accepts only ld-option elements as children, but found a "option" element.'
  //       )
  //     )
  //   }
  // })

  // it('throws if it can not parse custom Tether options', async () => {
  //   expect.assertions(1)
  //   try {
  //     const page = await newSpecPage({
  //       components: [LdSelect],
  //       html: `
  //         <ld-select placeholder="Pick a fruit" name="fruit" tetherOptions="asdf">
  //           <ld-option value="apple">Apple</ld-option>
  //           <ld-option value="banana">Banana</ld-option>
  //         </ld-select>
  //       `,
  //     })
  //
  //     await triggerPopper(page)
  //   } catch (err) {
  //     expect(err.toString()).toEqual(
  //       'ld-select failed parsing custom Tether options with JSON.parse.'
  //     )
  //   }
  // })

  it('applies size prop', async () => {
    const page = await newSpecPage({
      components: [LdSelect],
      html: `
        <ld-select placeholder="Pick a fruit" name="fruit" size="sm">
          <ld-option value="apple">Apple</ld-option>
          <ld-option value="banana">Banana</ld-option>
        </ld-select>
      `,
    })

    await new Promise((resolve) => setTimeout(resolve))

    const ldSelect = page.root
    expect(ldSelect.classList.contains('ld-select--sm')).toBeTruthy()

    await triggerPopper(page)

    const body = page.body
    const popper = body.querySelector('.ld-select__popper')
    expect(popper.classList.contains('ld-select__popper--sm')).toBeTruthy()
  })

  it('emits focus and blur event', async () => {
    const page = await newSpecPage({
      components: [LdSelect],
      html: `
        <ld-select placeholder="Pick a fruit" name="fruit" size="sm">
          <ld-option value="apple">Apple</ld-option>
          <ld-option value="banana">Banana</ld-option>
        </ld-select>
      `,
    })

    const ldSelect = page.root
    const triggerButton = ldSelect.querySelector('.ld-select__btn-trigger')

    const handlers = {
      onFocus() {
        return
      },
      onBlur() {
        return
      },
    }

    const spyFocus = jest.spyOn(handlers, 'onFocus')
    ldSelect.addEventListener('focus', handlers.onFocus)
    await triggerButton.dispatchEvent(new Event('focus'))
    await new Promise((resolve) => setTimeout(resolve))
    expect(spyFocus).toHaveBeenCalledTimes(1)

    const spyBlur = jest.spyOn(handlers, 'onBlur')
    ldSelect.addEventListener('blur', handlers.onBlur)
    await triggerButton.dispatchEvent(new Event('blur', { bubbles: true }))
    await new Promise((resolve) => setTimeout(resolve))
    expect(spyBlur).toHaveBeenCalledTimes(1)
  })

  it('uses checkboxes on options in multiple mode', async () => {
    const page = await newSpecPage({
      components: [LdSelect],
      html: `
        <ld-select placeholder="Pick a fruit" name="fruit" multiple>
          <ld-option value="apple">Apple</ld-option>
          <ld-option value="banana">Banana</ld-option>
        </ld-select>
      `,
    })

    await triggerPopper(page)

    const body = page.body
    const popper = body.querySelector('.ld-select__popper')
    const internalOptions = popper.querySelectorAll('ld-option-internal')

    expect(internalOptions.length).toEqual(2)
    expect(internalOptions[0].getAttribute('checkbox')).toEqual('true')
    expect(internalOptions[1].getAttribute('checkbox')).toEqual('true')
  })

  it('passes down prop selected option prop to internal options', async () => {
    const page = await newSpecPage({
      components: [LdSelect],
      html: `
        <ld-select placeholder="Pick a fruit" name="fruit">
          <ld-option value="apple">Apple</ld-option>
          <ld-option value="banana" selected='true'>Banana</ld-option>
        </ld-select>
      `,
    })

    await triggerPopper(page)

    const body = page.body
    const popper = body.querySelector('.ld-select__popper')
    const internalOptions = popper.querySelectorAll('ld-option-internal')

    expect(internalOptions.length).toEqual(2)
    expect(internalOptions[0].getAttribute('selected')).toEqual(null)
    expect(internalOptions[1].getAttribute('selected')).toEqual('true')
  })

  it('passes down prop preventDeselection to internal options', async () => {
    const page = await newSpecPage({
      components: [LdSelect],
      html: `
        <ld-select placeholder="Pick a fruit" name="fruit" prevent-deselection>
          <ld-option value="apple">Apple</ld-option>
          <ld-option value="banana">Banana</ld-option>
        </ld-select>
      `,
    })

    await triggerPopper(page)

    const body = page.body
    const popper = body.querySelector('.ld-select__popper')
    const internalOptions = popper.querySelectorAll('ld-option-internal')

    expect(internalOptions.length).toEqual(2)
    expect(internalOptions[0].getAttribute('prevent-deselection')).toEqual(
      'true'
    )
    expect(internalOptions[1].getAttribute('prevent-deselection')).toEqual(
      'true'
    )
  })
})
