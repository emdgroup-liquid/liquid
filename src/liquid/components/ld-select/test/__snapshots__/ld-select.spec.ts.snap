// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-select creates hidden input field, if inside a form 1`] = `
<ld-select name="fruit" placeholder="Pick a fruit">
  <mock:shadow-root>
    <div class="ld-select" part="root">
      <slot name="hidden"></slot>
      <div class="ld-select__slot-container" part="slot-container">
        <slot></slot>
      </div>
      <div class="ld-select__select" part="select">
        <div aria-disabled="false" aria-expanded="false" aria-haspopup="listbox" aria-label="Pick a fruit" class="ld-select__btn-trigger" part="btn-trigger focusable" role="button" tabindex="0">
          <span class="ld-select__btn-trigger-text-wrapper" part="trigger-text-wrapper" title="Pick a fruit">
            <span class="ld-select__btn-trigger-text" part="trigger-text">
              Pick a fruit
            </span>
          </span>
          <slot name="icon"></slot>
          <svg class="ld-select__icon" part="trigger-icon" role="presentation" viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg">
            <path d="M3 6l5 4 5-4" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="3"></path>
          </svg>
        </div>
      </div>
      <ld-select-popper role="listbox" style="z-index: 2147483646; display: none;">
        <mock:shadow-root>
          <div class="ld-select-popper" part="popper">
            <div class="ld-select-popper__scroll-container" part="popper-scroll-container">
              <slot></slot>
              <div class="ld-select-popper__shadow" part="shadow" style="height: 100%;"></div>
            </div>
          </div>
        </mock:shadow-root>
        <div part="options-container"></div>
      </ld-select-popper>
    </div>
  </mock:shadow-root>
  <ld-option class="ld-option" value="apple">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    Apple
  </ld-option>
  <ld-option class="ld-option" value="banana">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    Banana
  </ld-option>
  <input name="fruit" slot="hidden" type="hidden">
</ld-select>
`;

exports[`ld-select fills hidden input fields with initially selected option values 1`] = `
<ld-select multiple="" name="fruit" placeholder="Pick a fruit">
  <mock:shadow-root>
    <div class="ld-select" part="root">
      <slot name="hidden"></slot>
      <div class="ld-select__slot-container" part="slot-container">
        <slot></slot>
      </div>
      <div class="ld-select__select" part="select">
        <div aria-disabled="false" aria-expanded="false" aria-haspopup="listbox" aria-label="Pick a fruit" class="ld-select__btn-trigger" part="btn-trigger focusable" role="button" tabindex="0">
          <span class="ld-select__btn-trigger-text-wrapper" part="trigger-text-wrapper" title="Pick a fruit">
            <span class="ld-select__btn-trigger-text" part="trigger-text">
              Pick a fruit
            </span>
          </span>
          <slot name="icon"></slot>
          <svg class="ld-select__icon" part="trigger-icon" role="presentation" viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg">
            <path d="M3 6l5 4 5-4" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="3"></path>
          </svg>
        </div>
      </div>
      <ld-select-popper role="listbox" style="z-index: 2147483646; display: none;">
        <mock:shadow-root>
          <div class="ld-select-popper" part="popper">
            <div class="ld-select-popper__scroll-container" part="popper-scroll-container">
              <slot></slot>
              <div class="ld-select-popper__shadow" part="shadow" style="height: 100%;"></div>
            </div>
          </div>
        </mock:shadow-root>
        <div part="options-container"></div>
      </ld-select-popper>
    </div>
  </mock:shadow-root>
  <ld-option class="ld-option" selected="" value="apple">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    Apple
  </ld-option>
  <ld-option class="ld-option" selected="" value="banana">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    Banana
  </ld-option>
  <input name="fruit" slot="hidden" type="hidden" value="apple">
  <input name="fruit" slot="hidden" type="hidden" value="banana">
</ld-select>
`;

exports[`ld-select updates hidden input fields 1`] = `
<ld-select multiple="" name="fruit" placeholder="Pick a fruit">
  <mock:shadow-root>
    <div class="ld-select ld-select--expanded" part="root" style="z-index: 2147483647;">
      <slot name="hidden"></slot>
      <div class="ld-select__slot-container" part="slot-container">
        <slot></slot>
      </div>
      <div class="ld-select__select ld-tether-element-attached-left ld-tether-element-attached-top ld-tether-enabled ld-tether-target ld-tether-target-attached-bottom ld-tether-target-attached-left" part="select">
        <div aria-disabled="false" aria-expanded="true" aria-haspopup="listbox" aria-label="Pick a fruit" class="ld-select__btn-trigger" part="btn-trigger focusable" role="button" tabindex="0">
          <span class="ld-select__btn-trigger-text-wrapper" part="trigger-text-wrapper" title="Pick a fruit">
            <span class="ld-select__btn-trigger-text" part="trigger-text">
              Pick a fruit
            </span>
          </span>
          <slot name="icon"></slot>
          <svg class="ld-select__icon ld-select__icon--rotated" part="trigger-icon" role="presentation" viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg">
            <path d="M3 6l5 4 5-4" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="3"></path>
          </svg>
        </div>
        <div class="ld-tether-target-marker">
          <div class="ld-tether-marker-dot" style="top: 0px; left: 0px;"></div>
        </div>
      </div>
    </div>
  </mock:shadow-root>
  <ld-option class="ld-option" value="apple">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    Apple
  </ld-option>
  <ld-option class="ld-option" value="pear">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    Pear
  </ld-option>
  <ld-option class="ld-option" value="banana">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    Banana
  </ld-option>
  <input name="fruit" slot="hidden" type="hidden" value="pear">
  <input name="fruit" slot="hidden" type="hidden" value="banana">
</ld-select>
`;
